cmd/clusterctl/client/cluster/mover.go

```diff
@@ -592,6 +778,134 @@ func (o *objectMover) createTargetObject(nodeToCreate *node, toProxy Proxy) erro
+		Namespace: nodeToCreate.identity.Namespace,
+		Name:      nodeToCreate.identity.Name,
+	}
+
+	if err := cFrom.Get(ctx, objKey, obj); err != nil {
+		return errors.Wrapf(err, "error reading %q %s/%s",
+			obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
+	}
+
+	// Get JSON for object and write it into the configured directory
+	byObj, err := obj.MarshalJSON()
+	if err != nil {
+		return err
+	}
+
+	filenameObj := nodeToCreate.getFilename()
+	objectFile := filepath.Join(directory, filenameObj)
+
+	// If file exists, then remove it to be written again
+	if _, err = os.Stat(objectFile); err == nil {
+		if err = os.Remove(objectFile); err != nil {
+			return err
+		}
+	}
+
+	err = ioutil.WriteFile(objectFile, byObj, 0644)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *objectMover) restoreTargetObject(nodeToCreate *node, toProxy Proxy) error {
+	log := logf.Log
+	log.V(1).Info("Restoring", nodeToCreate.identity.Kind, nodeToCreate.identity.Name, "Namespace", nodeToCreate.identity.Namespace)
+
+	// Creates the targetObj into the target management cluster.
+	cTo, err := toProxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	// Attempt to retrieve an existing object. If it exists, update the UID to rebuild the owner chain
+	objKey := client.ObjectKey{
+		Namespace: nodeToCreate.identity.Namespace,
+		Name:      nodeToCreate.identity.Name,
+	}
+
+	existingTargetObj := &unstructured.Unstructured{}
+	existingTargetObj.SetAPIVersion(nodeToCreate.restoreObject.GetAPIVersion())
+	existingTargetObj.SetKind(nodeToCreate.restoreObject.GetKind())
+	if err := cTo.Get(ctx, objKey, existingTargetObj); err == nil {
+		log.V(5).Info("Object already exists, skipping restore", nodeToCreate.identity.Kind, nodeToCreate.identity.Name, "Namespace", nodeToCreate.identity.Namespace)
+
+		// Update the nodes UID since it already exits. Any nodes owned by this existing node will be updated when the owner chain is rebuilt
+		nodeToCreate.newUID = existingTargetObj.GetUID()
+
+		// Return early since the object already exists
+		return nil
+	}
+
+	// Rebuild the source object
+	obj := nodeToCreate.restoreObject
+
+	obj.SetAPIVersion(nodeToCreate.identity.APIVersion)
+	obj.SetKind(nodeToCreate.identity.Kind)
+
+	// New objects cannot have a specified resource version. Clear it out.
+	obj.SetResourceVersion("")
+
+	// Removes current OwnerReferences
+	obj.SetOwnerReferences(nil)
+
+	// Rebuild the owner reference chain
+	o.buildOwnerChain(obj, nodeToCreate)
+
+	if err := cTo.Create(ctx, obj); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "error creating %q %s/%s",
+				obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
+		}
+	}
+
+	// Stores the newUID assigned to the newly created object.
+	nodeToCreate.newUID = obj.GetUID()
+
+	return nil
+}
+
+// Recreate all the OwnerReferences using the newUID of the owner nodes.
+func (o *objectMover) buildOwnerChain(obj *unstructured.Unstructured, n *node) {
```

Comment:
does this also handle hierarchical owners? (obj A.ownerRef = obj B and obj B.ownerRef = obj C)

Reply:
I can't speak super well to the functionality of what this did before I refactored it into it's own function, but yes it should. This was simply yanked out of `createTargetObject` since it was duplicate code between that function, `restoreTargetObject`, and `backupTargetObject`. This can be found here: https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/client/cluster/mover.go#L531-L550

Once obj A is processed, it should have it's owner ref as obj B. Once obj B is processed, then it should then have it's owner obj as C. So, one by one, it should rebuild the entire owner tree.